service/worker/scheduler/activities.go

```diff
@@ -326,3 +283,81 @@ func translateError(err error, msgPrefix string) error {
 	}
 	return temporal.NewNonRetryableApplicationError(message, errType(err), err)
 }
+
+type responseBuilder struct {
+	request        *schedspb.WatchWorkflowRequest
+	workflowStatus enumspb.WorkflowExecutionStatus
+	logger         log.Logger
+}
+
+func newResponseBuilder(
+	request *schedspb.WatchWorkflowRequest,
+	workflowStatus enumspb.WorkflowExecutionStatus,
+	logger log.Logger,
+) responseBuilder {
+	return responseBuilder{request: request, workflowStatus: workflowStatus, logger: logger}
+}
+
+func (r responseBuilder) Build(event *historypb.HistoryEvent) (*schedspb.WatchWorkflowResponse, error) {
+	switch r.workflowStatus {
+	case enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING:
+		if r.request.LongPoll {
+			return nil, errTryAgain // not closed yet, just try again
+		}
+		return r.makeResponse(nil, nil), nil
+	case enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED:
+		if attrs := event.GetWorkflowExecutionCompletedEventAttributes(); attrs == nil {
+			return nil, errNoAttrs
+		} else if len(attrs.NewExecutionRunId) > 0 {
+			// this shouldn't happen because we don't allow old-cron workflows as scheduled, but follow it anyway
+			return nil, errFollow(attrs.NewExecutionRunId)
+		} else {
+			result := attrs.Result
+			if proto.Size(result) > maxResultSize {
```

Comment:
This condition can't happen, the workflow completion would have triggered an error in that case.

The problem is that if the response payload is e.g. (limit – 10 bytes), then the response payload of _this_ activity will be over the limit, because it gets wrapped in WatchWorkflowResponse, and then various local activity marker stuff. So we need to discard it if it's close to the limit. I suggest 1000 bytes as the assumed overhead size.